File: java/500/LeetCode0572.java
public class LeetCode0572 {
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (t == null) {
            return true;
        } else if (s == null) {
            return false;
        } else if (s.val == t.val && isSame(s, t)) {
            return true;
        } else {
            return isSubtree(s.left, t) || isSubtree(s.right, t);
        }
    }

    private boolean isSame(TreeNode t1,TreeNode t2) {
        if (t1 == null && t2 == null) {
            return true;
        } else if (t1 != null && t2 != null) {
            return t1.val == t2.val && isSame(t1.left, t2.left) && isSame(t1.right, t2.right);
        } else {
            return false;
        }
    }
}

File: cpp/L500/LeetCode0572.cc
#include "../heads/tree_node.h"
#include <queue>

class Solution {
    bool sameTree(TreeNode *rootA, TreeNode *rootB) {
        if (rootA == nullptr && rootB == nullptr) {
            return true;
        }
        if (rootA == nullptr && rootB != nullptr || rootA != nullptr && rootB == nullptr) {
            return false;
        }
        return rootA->val == rootB->val 
           && sameTree(rootA->left, rootB->left) 
           && sameTree(rootA->right, rootB->right);
    }
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (subRoot == nullptr) {
            return true;
        }
        if (root == nullptr) {
            return false;
        }
        std::queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            if (node->val == subRoot->val && sameTree(node, subRoot)) {
                return true;
            }
            if (node->left != nullptr) {
                q.push(node->left);
            }
            if (node->right != nullptr) {
                q.push(node->right);
            }
        }
        return false;
    }
};